### ISSUE TITLE
Error (Ecto.Query.CompileError) unknown key `:update` in update

### ISSUE THREAD

**Issue by jung-hunsoo** [elixir-ecto/ecto#2887]
I followed the doc [Ecto.Repo.update_all/3](https://hexdocs.pm/ecto/Ecto.Repo.html#c:update_all/3), but it throws an error when I included `:set` and `:inc` in a keyword list of `:update`. No error occurred when the two manipulations are separated.

### Environment
(dev env.)
* Elixir version (elixir -v): 1.7.0
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): PgSQL 9.4
* Ecto version (mix deps): 3.0.6
* Database adapter and version (mix deps): postgrex 0.14.1
* Operating system: ubuntu 14.04

### Current behavior

Include code samples, errors and stacktraces if appropriate.

Code snippet:
```elixir
  defp update_parent_counter(changeset) do
    changeset
    |> prepare_changes(fn pc ->
      case pc.action do
        :insert ->
          from(b in Brand, where: b.id == ^new_brand_id) |> pc.repo.update_all(inc: [products_count: 1])
          from(c in Category, where: c.id == ^new_category_id) |> pc.repo.update_all(inc: [products_count: 1])
          from(s in Segment, where: s.id == ^new_segment_id) |> pc.repo.update_all(inc: [products_count: 1])
          from(u in Userinfo, where: u.user_id == ^updated_by_id, update: [set: [last_new_tip_product_at: ^utc_now], inc: [tip_products_count: 1]]) |> pc.repo.update_all([]) # Error here
...
```

Error message:
```
Ecto.Query.CompileError at POST /tips/products
unknown key `:update` in update
```

Stacktrace:
```
Request: POST /tips/products
** (exit) an exception was raised:
    ** (Ecto.Query.CompileError) unknown key `:update` in update
        (ecto) lib/ecto/repo/queryable.ex:86: Ecto.Repo.Queryable.update_all/4
        (nohogu) lib/nohogu/tips/product.ex:151: anonymous fn/1 in Nohogu.Tips.Product.update_parent_counter/1
        (ecto) lib/ecto/repo/schema.ex:525: anonymous fn/2 in Ecto.Repo.Schema.run_prepare/2
        (elixir) lib/enum.ex:1925: Enum."-reduce/3-lists^foldl/2-0-"/3
        (ecto) lib/ecto/repo/schema.ex:241: anonymous fn/15 in Ecto.Repo.Schema.do_insert/3
        (ecto) lib/ecto/repo/schema.ex:917: anonymous fn/3 in Ecto.Repo.Schema.wrap_in_transaction/6
        (ecto_sql) lib/ecto/adapters/sql.ex:814: anonymous fn/3 in Ecto.Adapters.SQL.checkout_or_transaction/4
        (db_connection) lib/db_connection.ex:1349: DBConnection.run_transaction/4
        (nohogu) lib/nohogu_web/controllers/tips/product_controller.ex:50: NohoguWeb.Tips.ProductController.create/2
        (nohogu) lib/nohogu_web/controllers/tips/product_controller.ex:1: NohoguWeb.Tips.ProductController.action/2
        (nohogu) lib/nohogu_web/controllers/tips/product_controller.ex:1: NohoguWeb.Tips.ProductController.phoenix_controller_pipeline/2
        (nohogu) lib/nohogu_web/endpoint.ex:1: NohoguWeb.Endpoint.instrument/4
        (phoenix) lib/phoenix/router.ex:275: Phoenix.Router.__call__/1
        (nohogu) lib/nohogu_web/endpoint.ex:1: NohoguWeb.Endpoint.plug_builder_call/2
```

### Expected behavior

Set `:last_new_tip_product_at` to current utc and increase `:tip_produts_count` by 1.

**Comment by josevalim** [comment]
Just to double check, what if you call update_all(), without passing an
empty list?
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by jung-hunsoo** [comment]
@josevalim Well, simply errors.
```
UndefinedFunctionError at POST /tips/products
function Nohogu.Repo.update_all/1 is undefined or private
```

**Comment by josevalim** [comment]
@jung-hunsoo oh, yes. :D 

So I looked deeper into this and I cannot reproduce it. If you look [at this test case](https://github.com/elixir-ecto/ecto/blob/master/integration_test/cases/repo.exs#L980), I changed it to do both set and inc, and it works just fine.

Can you provide a minimal app that reproduces the issue?

**Comment by jung-hunsoo** [comment]
@josevalim Thanks.

Can you please check this code is working too?
`query = from p in Post, where: not is_nil(p.id), update: [set: [title: "2"], inc: [visits: 2]]`

I edited the test case as follows and ran the `mix test`, but no message nor error occurred.
```elixir
    # Positive
    IO.inspect TestRepo.get(Post, id1)
    query = from p in Post, where: not is_nil(p.id), update: [set: [title: "12"], inc: [visits: 2]]
    assert {2, nil} = TestRepo.update_all(query, [])
    IO.inspect TestRepo.get(Post, id1)
```

I followed the guide to run the test [Running tests](https://github.com/elixir-ecto/ecto.git), but it doesn't seem to run the cases under `integration_test`. `75 doctests, 894 tests, 0 failures` remains same after I removed a test case file there, too.

And when I ran the `mix test integration_test/cases/repo.exs`, it throws a compilation error.
```
== Compilation error in file integration_test/cases/repo.exs ==
** (CompileError) integration_test/cases/repo.exs:4: module Ecto.Integration.Case is not loaded and could not be found
    (elixir) expanding macro: Kernel.use/2
    integration_test/cases/repo.exs:4: Ecto.Integration.RepoTest (module)
```

**Comment by josevalim** [comment]
> query = from p in Post, where: not is_nil(p.id), update: [set: [title: "2"], inc: [visits: 2]]

Yup, it works too.

> I followed the guide to run the test Running tests, but it doesn't seem to run the cases under integration_test. 75 doctests, 894 tests, 0 failures remains same after I removed a test case file there, too.

To run the integration cases you need to run them from Ecto.SQL. Clone ecto_sql and then you can do:

    ECTO_ADAPTER=pg mix test path/to/ecto/integration_test/cases/repo.exs

**Comment by josevalim** [comment]
Ping.

**Comment by jung-hunsoo** [comment]
Sry for late. Currently a walkaround is used so I didn't try more. Let me try again tomorrow.

**Comment by jung-hunsoo** [comment]
@josevalim Finally I found the running codes, quite simple.

```elixir
utc_now = DateTime.utc_now()
from(u in Userinfo, where: u.user_id == ^updated_by_id) |> pc.repo.update_all(set: [last_new_tip_product_at: utc_now], inc: [tip_products_count: 1])
```
calls db query properly;
```
[debug] QUERY OK source="general_userinfos" db=0.9ms
UPDATE "general_userinfos" AS g0 SET "last_new_tip_product_at" = $1, "tip_products_count" = g0."tip_products_count" + $2 WHERE (g0."user_id" = $3) [#DateTime<2019-02-06 18:08:50Z>, 1, 2]
```

This is the expected result.

**Comment by josevalim** [comment]
So is everything behaving as expected in our suite? I.e. you cannot reproduce the issue?

**Comment by jung-hunsoo** [comment]
@josevalim Yep. The original code also works as expected. I don't know why it didn't work. It's assumed that the code changes weren't compiled lively for some unknown reason.

In addition the two codes create different query interestingly. (not a problem)

```elixir
from(u in Userinfo, where: u.user_id == ^updated_by_id) |> pc.repo.update_all(set: [last_new_tip_product_at: utc_now], inc: [tip_products_count: 1])
```
produces
```elixir
UPDATE "general_userinfos" AS g0 SET "last_new_tip_product_at" = $1, "tip_products_count" = g0."tip_products_count" + $2 WHERE (g0."user_id" = $3) [#DateTime<2019-02-06 18:08:50Z>, 1, 2]
```
while
```elixir
from(u in Userinfo, where: u.user_id == ^updated_by_id, update: [set: [last_new_tip_product_at: ^utc_now], inc: [tip_products_count: 1]]) |> pc.repo.update_all([])
```
produces
```elixir
UPDATE "general_userinfos" AS g0 SET "tip_products_count" = g0."tip_products_count" + 1, "last_new_tip_product_at" = $1 WHERE (g0."user_id" = $2) [#DateTime<2019-02-0618:28:32Z>, 2]
```

**Comment by josevalim** [comment]
Awesome, glad to know it is all working as expected!
